#!/usr/bin/env Rscript
#args = commandArgs(trailingOnly=TRUE)

options(repos=structure(c(CRAN="https://cran.cnr.berkeley.edu/")))

help_text="
This is pcapred.R\
Usage: Rscript pcapred.R <options>\
pcapred.R maps external data into the PCA space generated by flashpca or plink.\
Author: Daniel Lawson: dan.lawson@bristol.ac.uk\
Version: 1.0.1\
Copyright University of Bristol 2019. All rights reserved.\
"


#######################################################
#######################################################
#######################################################
#######################################################

packages=c("optparse","data.table","R.utils","pcapred")
test=sapply(packages,function(x){
    try(suppressMessages(require(x,character.only=T)),silent=T)
})
    
if(any(!test)){
    cat(help_text)
    cat("IMPORTANT! To use this code you need to :
install.packages(\"optparse\")
install.packages(\"data.table\")
install.packages(\"R.utils\")
remotes::install_github(\"danjlawson/pcapred\")

Attempting to do that now.")
    
    install.packages("optparse")
    install.packages("data.table")
    install.packages("R.utils")
    
    test=sapply(packages,function(x){
        try(suppressMessages(require(x,character.only=T)),silent=T)
    })
    if(any(!test)){
        stop("Packages required; installation failed. Do this manually.")
    }else{
        cat("Installed correctly; continuing.")
    }
}
source("/Users/madjl/playground/pcapred/pcapred/R/input.R")
source("/Users/madjl/playground/pcapred/pcapred/R/mappingfunctions.R")
source("/Users/madjl/playground/pcapred/pcapred/R/output.R")

option_list = list(
    make_option(c("-f", "--file"), type="character", default=NULL, 
                help="plink-format file root (with endings .bim/.bed/.fam)"),
    make_option(c("-r", "--ref"), type="character", default=NULL, 
                help="reference data package:
.afreq.gz: as output by plink2 with --freq and gzipped
THEN if mode==plink:
.eigenval/.eigenvec.var.gz: as output by plink2 with --pca var-wts and .var file (var is gzipped)
OR if mode==flashpca:
.val/.load.gz: as output by flashpca with --outval and --outload (load is gzipped)"),
    make_option(c("-T", "--transpose"),default=FALSE,
                help="Transposed input weights. Named .tvar.gz or .tload.gz.
Will allow lower memory requirements in the future.", action="store_true"),
    make_option(c("-o", "--out"), type="character", default=NULL, 
                help="output file name [default=<file>.eigenvec.pred]"),
    make_option(c("-m", "--minfreq"), type="numeric", default=0.001, 
                help="Minimum allele frequency [default=%default]"),
    make_option(c("-q","--quiet"),action="store_false",
          dest="verbose", help="Print little output"),
    make_option(c("-v","--verbose"), default=TRUE,
                help="Verbose mode (default: TRUE; use -q to quiet it)", action="store_true"),
    make_option(c("-M","--mode"), type="character",default="flashpca",
                help="Mode of the data. Options are flashpca or plink"),
    make_option(c("-b","--batchsize"), type="numeric",default="0",
                help="Process the input data in batches of this size. Default does all in one go."),
    make_option(c("-t","--test"), default=FALSE,
                help="Test mode: looks for a <file>.eigenvec (plink) or <file>.pcs (flashpca) file, reads it in, and does some tests to compare whether we are getting the same sort of results.", action="store_true")
); 
 
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
opt$raw=FALSE

help=function(){
    cat(help_text)
    print_help(opt_parser)
    stop("Options required.")
}

if(opt$help) {
    print("Requested help")
    help()
}

if(is.null(file) | is.null(opt$ref)){
    print("MUST PROVIDE --file and --ref")
    help()
}

#######################################################
#######################################################
#######################################################
#######################################################

## What we need in order to do the pca:
## Input data in plinks raw format, as output by --recode A
## Reference data in the form of plinks freq table, as output by --freq
## SNP weights, as output by plink by --pca var-wts (modified by transposing)
## eigenvalues, as output by plink by --pca (.eigenval)



out=opt$out
if(is.null(out)){
    out=gsub(".raw.gz",".eigenvec.pred",opt$file,fixed=T)
    out=gsub(".raw",".eigenvec.pred",out,fixed=T)
}
outfile=out
testfile=""
if(opt$test){
    if(opt$mode=="plink"){
        testfile=paste0(opt$file,".eigenvec")
#        testfile=gsub(".raw.gz",".eigenvec",opt$file,fixed=T)
#        testfile=gsub(".raw",".eigenvec",testfile,fixed=T)
    }else{
        testfile=paste0(opt$file,".pcs")
#        testfile=gsub(".raw.gz",".pcs",opt$file,fixed=T)
#        testfile=gsub(".raw",".pcs",testfile,fixed=T)
    }
}
if(opt$verbose){
    cat("Options in use:\n")
    cat(paste0("input plink file = ",opt$file,"\n"))
    cat(paste0("mode = ",opt$mode,"\n"))
    cat(paste0("ref = ",opt$ref,"\n"))
    ## cat(paste0("  : ",opt$mode," eigenval file = ",pcvalsfile,"\n"))
    ## cat(paste0("  : ",opt$mode," SNP weights file = ",snpweightsfile,"\n"))
    ## cat(paste0("  : plink allele frequency file = ",afreqfile,"\n"))
    cat(paste0("output file = ",out,"\n"))
    cat(paste0("minimum frequency = ",opt$minfreq,"\n"))
    cat(paste0("test = ",test,"\n"))
    if(opt$test) cat(paste0("testfile = ",testfile,"\n"))
}

###########################
###########################
###########################
## Test that the files exist BEFORE we waste time doing things

checkfiles=list(paste0(opt$file,".bed"),
               paste0(opt$file,".bim"),
               paste0(opt$file,".fam")
               )
if(opt$test)checkfiles=c(checkfiles,testfile)
testfiles(checkfiles,opt$verbose)
ref=readreference(opt$ref,mode=opt$mode,
                          transpose=opt$transpose,checkonly=TRUE,minfreq=opt$minfreq,verbose=opt$verbose)

###########################
###########################
###########################
#####################
if(opt$verbose) cat(paste("Reading Data\n"))

## Read the raw data
if(opt$verbose) cat(paste("... Reading SNP data of file to project: ",opt$file,"\n"))
dat=readinput(opt$file,opt$raw,opt$verbose)

## Read the eigenvalues
if(opt$verbose) cat(paste("... Reading Reference data\n"))
ref=readreference(opt$ref,mode=opt$mode,
                  transpose=opt$transpose,
                  verbose=opt$verbose)

#####################
#####################
#####################
## Normalise the raw data and reference panel
if(opt$verbose) cat(paste("Standardizing the reference data and the reference panel\n"))

dat=mergeref(dat,ref,opt$verbose)

#####################
#####################

if(opt$verbose) cat(paste("Performing Computation\n"))

## Perform the inference
if(opt$verbose) cat(paste("... Computing predictions\n"))
pred=predictpcs(dat,verbose=opt$verbose)

## Write output
if(opt$verbose) cat(paste("Creating output\n"))
write_pred(outfile,dat$indinfo,pred)
cat(paste("Written predicted eigenvectors to",outfile,"\n"))

## Testing
if(opt$test){
    if(opt$verbose) cat(paste("Testing: Performing comparisons with",testfile,"\n"))
    testres=test_pca(testfile,pred,outfile,opt$verbose)    
}

